Add RemoveFinalReset pass

garrison opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This adds a `RemoveFinalReset` pass, which removes a `Reset` instruction if it comes at the end of a qubit wire.  ~~It also enables this pass at optimization level 1 and higher.~~

### Details and comments

The new wire cutting code in the Circuit Knitting Toolbox results in circuits where there are `Reset`s in the circuit.  In the special case that no qubits are re-used later in the circuit, then all of these `Reset`s are either at the beginning of the circuit or the end of the circuit.  There is an existing pass, `RemoveResetInZeroState`, that removes resets at the beginning of the circuit.  This pass is disabled by default in the pass managers (see https://github.com/Qiskit/qiskit/pull/10591), as it places assumptions on the state at the beginning of the circuit.  This PR creates the other transpiler pass that we need, namely one that removes resets at the _end_ of the circuit.  If nothing happens to a qubit wire after, then this removal should have no effect on the circuit.  Of course, this _does_ change the result if you then concatenate it with another circuit; ~~I'm assuming this assumption is okay, but if not, it should be removed from the default passes.~~ EDIT: I have removed this from the default passes in 4adc04889f76239e475a2ddd33929fb0d4cf267a.

Related CKT issue: https://github.com/Qiskit-Extensions/circuit-knitting-toolbox/issues/452

garrison committed:
Add RemoveFinalReset pass

mtreinish commented on qiskit/transpiler/passes/optimization/remove_final_reset.py:
```diff
@@ -0,0 +1,37 @@
+        for reset in resets:
+            successor = next(dag.successors(reset))
+            if isinstance(successor, DAGOutNode):
+                dag.remove_op_node(reset)
```
It'll probably be a bit more efficient to use the output_map in the dag and look at it's predecessors. This is because `dag.op_nodes()` iterates over all the instructions in the dag and in a typical circuit there will be far more instructions than qubits and clbits.

```suggestion
        for output_node in dag.output_map.values():
            if isinstance(output_node.wire, Qubit):
                pred = next(dag.predecessors(output_node))
                if isinstance(pred, DAGOpNode) and isinstance(pred.op, Reset):
                    dag.remove_op_node(pred)
```

garrison replied:
Fixed in f358d46c33925f68598b67151caa1b8ba2de5631. I wonder if this could potentially be improved in the `RemoveResetInZeroState` pass too (I modeled my original code on this pass).

mtreinish replied:
Yeah, we can probably make a similar change to `RemoveResetInZeroState`. It looks like it hasn't had any meaningful updates since it was first introduced in #2139.

garrison committed:
Fix lint

garrison committed:
Look at predecessors of `output_map`

Suggested at https://github.com/Qiskit/qiskit/pull/11266#discussion_r1397572149

Co-authored-by: Matthew Treinish <mtreinish@kortar.org>

garrison committed:
Remove `RemoveFinalReset` from preset passmanagers

This pass modifies the quantum output state of the circuit, so it
should not be run by default.

mtreinish committed:
Merge branch 'main' into remove-final-reset

garrison committed:
Update remove-final-reset-488247c01c4e147d.yaml

It's no longer run by default at any optimization level.

garrison committed:
Fix `QiskitTestCase` import

garrison committed:
Update to use `DoWhileController`

garrison commented:
@mtreinish: I fixed a few things; this is ready to merge now.

mtreinish commented on test/python/transpiler/test_remove_final_reset.py:
```diff
@@ -15,10 +15,11 @@
 from qiskit.transpiler.passes import RemoveFinalReset, DAGFixedPoint
 from qiskit.converters import circuit_to_dag
-from qiskit.test import QiskitTestCase
+from test import QiskitTestCase  # pylint: disable=wrong-import-order
```
FWIW, you could make pylint happy if you added a new line between the `qiskit` import section and this import.
